From: Matthew Wilcox <willy@infradead.org>
Subject: mm-convert-ksm_might_need_to_copy-to-work-on-folios-fix
Date: Wed, 13 Dec 2023 15:37:25 +0000

The patch 6faef19bdfcc: "mm: convert ksm_might_need_to_copy() to work
on folios" from Dec 11, 2023 (linux-next), leads to the following
Smatch static checker warning:

 	mm/memory.c:4118 do_swap_page()
 	error: 'folio' dereferencing possible ERR_PTR()

Link: https://lkml.kernel.org/r/ZXnPtblC6A1IkyAB@casper.infradead.org
Signed-off-by: Matthew Wilcox <willy@infradead.org>
Reported-by:  Dan Carpenter <dan.carpenter@linaro.org>
Closes: https://lkml.kernel.org/r/a07e5aad-2b09-4966-b486-4791619dceae@moroto.mountain
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/memory.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/mm/memory.c~mm-convert-ksm_might_need_to_copy-to-work-on-folios-fix
+++ a/mm/memory.c
@@ -3945,9 +3945,11 @@ vm_fault_t do_swap_page(struct vm_fault
 		folio = ksm_might_need_to_copy(folio, vma, vmf->address);
 		if (unlikely(!folio)) {
 			ret = VM_FAULT_OOM;
+			folio = swapcache;
 			goto out_page;
 		} else if (unlikely(folio == ERR_PTR(-EHWPOISON))) {
 			ret = VM_FAULT_HWPOISON;
+			folio = swapcache;
 			goto out_page;
 		}
 		page = folio_page(folio, 0);
_
